### ISSUE TITLE
getting the same prediction output after freezed the model

### ISSUE THREAD

**Issue by kerolos** [tflearn/tflearn#1063]
I saved the model as explained in the code below.
Here is my intput and output layer (names and types):
tf_input: Tensor("prefix/input/X:0", shape=(?, 64, 64, 3), dtype=float32)
tf_output: Tensor("prefix/FullyConnected_2/Softmax:0", shape=(?, 4), dtype=float32)

my question is when I run this freezed model in c++ I always get the same prediction result it does not change.it might be I made the input as a constant.
I really appreciate your help.

#### run the session in c++
string input_layer = "input/X:0";//"input/X:0";
string output_layer = "FullyConnected_2/Softmax:0";

Status status = m_session->Run({{input_layer,resized_tensor}},
            {output_layer}, {}, &outputs);
std::cout << outputs[0].DebugString() << "\n";   ----> the output is constant for all images

### Freeze Function 
def freeze_graph(model):
    # We precise the file fullname of our freezed graph
    #output_graph = "/mnt/data_disk/ExternalHD3/FlappenDataset/SnaecktakulaerSmall2018/result/frozen_model.pb"

    # Before exporting our graph, we need to precise what is our output node
    # This is how TF decides what part of the Graph he has to keep and what part it can dump
    # NOTE: this variable is plural, because you can have multiple output nodes
    output_node_names = "FullyConnected_2/Softmax"#  "FullyConnected_2" #input/X,

    # We clear devices to allow TensorFlow to control on which device it will load operations
    clear_devices = True
    # We import the meta graph and retrieve a Saver
    #saver = tf.train.import_meta_graph(input_checkpoint + '.meta', clear_devices=clear_devices)

    # We retrieve the protobuf graph definition
    graph = model.net.graph
    #input_graph_def = graph.as_graph_def()

    # We start a session and restore the graph weights
    # We use a built-in TF helper to export variables to constants
    sess = model.session

    for v in sess.graph.get_operations():
      print(v.name)
    # for fixing the bug of batch norm
    #gd = sess.graph.as_graph_def()

    #with sess.graph.as_graph_def():
     # del tf.get_collection_ref(tf.GraphKeys.TRAIN_OPS)[:]

    gd = sess.graph.as_graph_def()

    for node in gd.node:            
        if node.op == 'RefSwitch':
            node.op = 'Switch'
            for index in range(len(node.input)):
                if 'moving_' in node.input[index]:
                    node.input[index] = node.input[index] + '/read'
        elif node.op == 'AssignSub':
            node.op = 'Sub'
            if 'use_locking' in node.attr: del node.attr['use_locking']
        elif node.op == 'AssignAdd':
            node.op = 'Add'
            if 'use_locking' in node.attr: del node.attr['use_locking']

    output_graph_def = graph_util.convert_variables_to_constants(
        sess, # The session is used to retrieve the weights
        #input_graph_def
        gd, # The graph_def is used to retrieve the nodes 
        output_node_names.split(",") # The output node names are used to select the usefull nodes
    ) 

    # Finally we serialize and dump the output graph to the filesystem
    with tf.gfile.GFile(output_graph, "wb") as f:
        f.write(output_graph_def.SerializeToString())
    print("%d ops in the final graph." % len(output_graph_def.node))

**Comment by kerolos** [comment]
furthermore, I tried to use python code for classifying an image and I got a wrong result. 

##Python code: 
import tensorflow as tf
import sys
import os

# speicherorte fuer trainierten graph und labels in train.sh festlegen ##

# Disable tensorflow compilation warnings
os.environ['TF_CPP_MIN_LOG_LEVEL']='2'
import tensorflow as tf

image_path = sys.argv[1]
# angabe in console als argument nach dem aufruf  


#bilddatei readen
image_data = tf.gfile.FastGFile(image_path, 'rb').read()

# holt labels aus file in array 
label_lines = [line.rstrip() for line 
                   in tf.gfile.GFile("tf_files/retrained_labels.txt")]
# !! labels befinden sich jeweils in eigenen lines -> keine aenderung in retrain.py noetig -> falsche darstellung im windows editor !!
				   
# graph einlesen, wurde in train.sh -> call retrain.py trainiert
with tf.gfile.FastGFile("tf_files/retrained_graph.pb", 'rb') as f:
 
    graph_def = tf.GraphDef()	## The graph-graph_def is a saved copy of a TensorFlow graph; objektinitialisierung
    graph_def.ParseFromString(f.read())	#Parse serialized protocol buffer data into variable
    _ = tf.import_graph_def(graph_def, name='')	# import a serialized TensorFlow GraphDef protocol buffer, extract objects in the GraphDef as tf.Tensor
	
	#https://github.com/Hvass-Labs/TensorFlow-Tutorials/blob/master/inception.py ; ab zeile 276

with tf.Session() as sess:

    softmax_tensor = sess.graph.get_tensor_by_name('final_result:0')
	# return: Tensor("final_result:0", shape=(?, 4), dtype=float32); stringname definiert in retrain.py, zeile 1064 

    predictions = sess.run(softmax_tensor, \
             {'DecodeJpeg/contents:0': image_data})
    # gibt prediction values in array zuerueck:
	
    top_k = predictions[0].argsort()[-len(predictions[0]):][::-1]
	# sortierung; circle -> 0, plus -> 1, square -> 2, triangle -> 3; array return bsp [3 1 2 0] -> sortiert nach groesster uebereinstimmmung

	# output
    for node_id in top_k:
        human_string = label_lines[node_id]
        score = predictions[0][node_id]
        print('%s (score = %.5f)' % (human_string, score))

**Comment by kerolos** [comment]
furthermore, when i tryed to classify the trained model as a format (pb), it gives unpredicatable result.

### python code:
  
import tensorflow as tf
import sys
import os
os.environ['TF_CPP_MIN_LOG_LEVEL']='2'

image_path = sys.argv[1]
image_data = tf.gfile.FastGFile(image_path, 'rb').read()

label_lines = [line.rstrip() for line 
                   in tf.gfile.GFile("tf_files/retrained_labels.txt")]
with tf.gfile.FastGFile("tf_files/retrained_graph.pb", 'rb') as f:
 
    graph_def = tf.GraphDef() 
    graph_def.ParseFromString(f.read()) 
    _ = tf.import_graph_def(graph_def, name='') 
with tf.Session() as sess:

    softmax_tensor = sess.graph.get_tensor_by_name('FullyConnected_2/Softmax:0')

    predictions = sess.run(softmax_tensor, \
             {'input/X:0': image_data})
   
    top_k = predictions[0].argsort()[-len(predictions[0]):][::-1]

    for node_id in top_k:
        human_string = label_lines[node_id]
        score = predictions[0][node_id]
        print('%s (score = %.5f)' % (human_string, score))

 ### Out put:

c7J\xf5\xbf\xd9\xe6xc\xb8\xd5\xe4\x9c\x1d\xcbh6\xe3\xd75\x86c\x1el4\x91\xbd\x17y\xa3\xba\xf1+\xad\xd6\xa0\xeaXmL\xe0\x81\xcdr\xfa\xde\x98\x97\x114\xc9o\xed\xbb\xd6\xac\xebw\xbe*\xb5)q\xa3iIy\xb9\x8e\xf4\x92M\xbbG\xd6\xb1\xed5\xdf\x15\xde\\\xb4Z\x9f\x86\xd2\xd6-\xa4\xb3,\xe1\xc6}+\xc8\xa3N\xa4"\xa4\x9a\xd0\xeb\x9c\x93\xbcY\xf1o\xc7}\x0e\xefB\xf8\x83\xa8Y\\\xc2Su\xcb:\xf1\xfc\'\xa5q%O@\xbd\xb8\x15\xf5\xdf\xc7\x8f\x81ZO\xc4\rb\xd7_\xbc\xbe\x16\xa28\xb6\xc8\x15@.3\xdc\xd7%\xf1+\xe1o\xc2o\x00\xfc)\x9ao\xec\xd8\x8d\xe3FD3;\xfe\xf1\x9f\xb6+\xef\xf0y\xde\x19\xd1\xa7\x0b7\'\xa5\x91\xf3\xf5\xb0\x15T\xa5\'\xb1\xe3\xba\x13H\xde\x1e\xb6R~Uf"\xbdk\xf6x\xf16\xa5c\xa3x\x9bB\x0c\xc6\xda]"I6\xe7 0\x18\xe9^M\xe1\xad\xc7H\xd8N0\xdd+\xdd~\x00|<\xd4\xed~\x15\xf8\x9f\xc7\xf76\xa5b}:H`\xcf\xf1z\x91]\x19\xb4\xe9\xc7\x0c\xf9\xfa\xb5\xf7\xdc\xe7\xc2FR\xac\xf9Mx\xe3\x0c\xcf\xe5\xa8\xdc\x0e9\x1dk\xd5\xff\x00gK\x0bk\xcb\xfdN\xd2S\x815\xb0\xc9\x1dx\xaf-\xb5V\x8a\xe7.\xa5\xb0\x0f$q^\xa1\xfb;\xc4d\xd4\xaf\xe7\x8e26\xc1\xd4\x1e\x95\xf1x\xe6\xde\x1aL\xf7\xe8\xbfz\xe7]\xa8@\xf6D\xc0\x8f\xf2\x92@e\x15\xccx\xa3Q:^\x9bsymn\xb24Q3\xacC\x8d\xcc\x05usi\xda\x87\x88\xf5_\xec\xad #>\x0b8\r\x8cc\xb9\xaam\xf0\xc3T\xd4.\xcc:\xc4\xe9\x1c\x1c\x89J\x1c\x92==\xeb\xc8\xa4\xe9\xc5\xa79/3\xae\\\xdc\x9a\x1f\x14|s\xfd\xa2\xfcW\xe3;\x98t\xcb\x10t\xf8\xa0\x18x\xe3s\x96`{\xd7\x9d\xf8\x83\xc6>)\xf1\xad\xd4Qj\x97\xf2\xdc\x15\xc2\xc3\x18\xcf\x1fA_L|B\xff\x00\x82|x\xb7\xc5^.\xb9\xd4|/\xafZCi+\x97H\xe7\xc8e\xad\xef\x83\xbf\xf0O\xc9\xbc1\xa9\xc7\xaa\xf8\x9fX\xb4\xba\xb8\x8d\xf2\x80/\x0b\x8e\x86\xbe\xf2\x9ei\x91`\xb0\xcaPj\xe9^\xd6\xd4\xf0^\x17\x1d^\xab\x8b\xbd\xaf\xf29\xff\x00\xd9\xc7\xf6,\xd6u\xff\x00\x0f[x\xbb\xe2\x04\xdfe\xb4\x90\x87\x86\xcf\xf8\xe4\x1e\xa7\xd0W\xbex\xff\x00\xc2\xab\xa7|.\xd4<\t\xe1+(\xe3g\xb2hm\xe3L(l\x8e\xb5\xdb\xc7bt\x9d\x1e\r-\xe6R\xb6i\xb7n1\x9a\xce\xb8\x96\x1b\xd7V\xc6@\x1cg\xa85\xf1\x18\xbc\xdf\x17\x98bUI\xbd"\xee\x97E\xea{\x9404\xe8G\x92=z\x9f\xff\xd9'
